File: tests/fixtures/wc.py
"""Reference wc implementation for golden task 1.

Deterministic library: count words, lines, characters, and bytes in a string.
This is the baseline that the Rust port must match byte-for-byte.
"""


def wc(text: str) -> dict[str, int]:
    """Count words, lines, characters, and bytes in a string.

    Args:
        text: Input string (any encoding).

    Returns:
        Dict with keys: words, lines, chars, bytes.
    """
    if not text:
        return {"words": 0, "lines": 0, "chars": 0, "bytes": 0}

    lines = text.count("\n")
    # Count trailing content after last newline as a line
    if text and not text.endswith("\n"):
        lines += 1
    elif not text.strip():
        # Edge case: only whitespace/newlines, treat as counted lines
        lines = text.count("\n")

    words = len(text.split())
    chars = len(text)
    b = text.encode("utf-8")

    return {
        "words": words,
        "lines": lines,
        "chars": chars,
        "bytes": len(b),
    }

File: mew-core/crates/engine/src/migration/wc.rs
//! Word/line/character/byte counter.
//!
//! Ported from `tests/fixtures/wc.py`. All behavior must match the Python
//! baseline byte-for-byte for every valid UTF-8 input, including edge cases
//! (empty string, whitespace-only, trailing newlines).

/// Count words, lines, Unicode scalar values, and UTF-8 bytes in `text`.
///
/// Returns `(words, lines, chars, bytes)`. Behavior matches the Python
/// `wc.py` baseline: `str::lines()` semantics for line counting,
/// `str::split_whitespace()` for word boundaries.
pub fn wc(text: &str) -> (usize, usize, usize, usize) {
    if text.is_empty() {
        return (0, 0, 0, 0);
    }

    let lines = text.lines().count();
    let words = text.split_whitespace().count();
    let chars = text.chars().count();
    let byte_count = text.len();

    (words, lines, chars, byte_count)
}
